Support OpenSSL 1.1.0

winksaville opened the pull request:
This resolves issue [#1838](https://github.com/ponylang/ponyc/issues/1838)

winksaville commented:
This should not be failing in Travis, the default is basically a "NOP" if OpenSSL 1.1 is not installed. Which I don't believe it is on and of the Travis enviornments. It does work for me on Arch Linux where OpenSSL 1.1 is installed.

In anycase it needs to be reviewed and I also need to create some tests, as such I'd like this to be marked with "Do not merge" for now.

EpicEric commented:
The Travis failures have to do with network issues with the jobs, so this is likely not the reason for these errors.

I'll still try to get around testing this.

SeanTAllen commented:
@jemc marking as needs discussion. want to talk this over with you. perhaps next sync.

winksaville commented:
SG, I'll make my best attempt to attend the next sync.

mfelsche commented:
Before we release this, we should add another part to our travis test runs, that tests the stdlib against openssl 1.1.0.

mfelsche commented:
The only downside of this approach is that it adds the burden on the application programmer to define "OPENSSL_API_COMPAT_1_1_0" for every compile run of his pony program.

It would be nice if the detection of which ssl library is available and would be picked up by FFI could be done in the compiler so we can use it in ifdef statements to chose the right function.

mfelsche commented:
We agreed on going with this approach for unlocking.
We would love to have one define called `openssl_1.1.0` that corresponds to the current `OPENSSL_API_COMPAT_1_1_0`.

The automatic detection of the openssl version does not work for every platform and it is only necessary in the Makefile rules that compile the stdlib for testing, thus it should not error if no version could be detected.

The idea is to require the user to add a `-Dopenssl_1.1.0` in case he wants to link against OpenSSL 1.1.0 on his platform and to also document this prominently both in the stdlib docs of the `net/ssl` and the `crypto` packages and in the tutorial. 

A failsafe automatic OpenSSL version detection for convenience should be subject of another PR and discussion.

Can we ask you to give the doc part a try? If not, I could chime in to help here.

winksaville commented:
I'll make the changes to the code today and follow with changes to the docs afterwards.

winksaville commented:
So a first pass at trying to add a compiler ifdef, I'm not sure I've got this right but its "working" for me. I tried to define openssl_1.1.0 but I got a syntax error so I switched to openssl_1_1_0:
```
Linking libponyrt.benchmarks
build/release/ponyc -DOPENSSL_1.1.0 --checktree --verify packages/stdlib
Building builtin -> /home/wink/prgs/pony/ponyc/packages/builtin
Error:
/home/wink/prgs/pony/ponyc/packages/builtin/platform.pony:20:16: syntax error: expected ( after openssl_1
  fun openssl_1.1.0(): Bool => compile_intrinsic
               ^
```

winksaville commented:
So at the moment openssl_1_1_0 is always defined, so I've done something totally wrong. Guidance appreciated.

mfelsche commented:
Sorry if i was misleading.
I actually meant, changing the ifdef string value to "openssl_1.1.0" and set it accordingly in the makefile as a ponyc define with "-D" as you do right now with "OPENSSL_API_COMPAT_1_1_0". but only add it as define in the makefile if you detected openssl 1.1.0 and otherwise add no define.

winksaville commented:
Ok, so as I understand it, the ifdef in the code would remain a string but
would be defined as:
''''
    -D "openssl_1.1.0"
'''''
And tested with:
'''
   Ifdef "openssl_1.1.0" then
     foo..
   else
     bar..
   end
'''

On Thu, Dec 28, 2017, 11:54 PM Matthias Wahl <notifications@github.com>
wrote:

> Sorry if i was misleading.
> I actually meant, changing the ifdef string value to "openssl_1.1.0" and
> set it accordingly in the makefile as a ponyc define with "-D" as you do
> right now with "OPENSSL_API_COMPAT_1_1_0". but only add it as define in the
> makefile if you detected openssl 1.1.0 and otherwise add no define.
>
> —
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ponylang/ponyc/pull/2415#issuecomment-354411879>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA-hHMosKWOt2yF0h5_MORIyHgSUscPhks5tFJrRgaJpZM4Q_4Lg>
> .
>

mfelsche commented:
correct.

winksaville commented:
@mfelsche, please see my latest changes and let me know what other changes you like to the code.

winksaville commented:
I've made changes to README.md and rebased on master, squashed the commits. As requested I've got a PR for the [FAQ page](https://github.com/ponylang/ponylang.github.io/pull/228) in the pony website.

dipinhora commented:
I'm late to the discussion on this, but it might be better to do something similar to how `make use=scheduler_scaling_pthreads` works where the `Makefile` adds a define/build flag for the C compiler (`-DUSE_SCHEDULER_SCALING_PTHREADS` for the `scheduler_scaling_pthreads` case) and then the code in `main.c` enables the appropriate pony level define/build flag to automagically be added to all future `ponyc` invocations (see https://github.com/ponylang/ponyc/blob/master/src/ponyc/main.c#L298-L302 for the specifics of the `scheduler_scaling_pthreads` case).

If we did this, it would be a relatively seamless/painless process for folks on distributions with openssl 1.1.0 to use where they would only have to specify something special once during the build of ponyc (like `use=scheduler_scaling_pthreads` and `default_pic=true` currently work) for compatibility with their environment and not have to remember something special for every invocation of `ponyc` itself.

winksaville commented:
I agree a better solution is needed. My first implemenation added code to the Makefile which detected the installed version and set the "OPENSSL" makefile variable automatically. That code was modeled after the LLVM detection, but people didn't like that so it morphed into what we have now.

This was dicussed, on the sync call and we reached a conclusion that something better was needed long term, but this was good enough for now as not alot of people are affected.

My opinion is that we shouldn't modify the compiler, the compiler doesn't care. If a package needs to support multiple variants of a library then I'd say the package should be obligated to determine what library version is "current".

Just thinking out loud, a package should have a compile time actor, maybe _compile.pony, and that interface that might have two methods:
  - Get "current" library version.
  - Set "current" library version, with the default being "best" version found.

dipinhora commented:
I agree the compiler shouldn't have to care about the version of `openssl` or even whether `pic` is required or not. However, given that the ssl support is part of a package that is bundled with the compiler muddies things. I believe programmer ergonomics need to be considered for something that is effectively a part of the compiler due to being packaged with it.

While I agree that not a lot of people are affected currently, we should keep in mind whether people want to use a tool or not is related to how easy/user friendly it is. In my opinion, the `default_pic=true` (and my suggested similar approach for this openssl version thing) is a relatively minor compromise with a large upside. Being able to compile with these things defaulted appropriately for a particular environment allows for distros to easily package ponyc for their library versions (and `--pic` requirements) without end users having to know anything about decisions the distro made.

SeanTAllen commented:
I agree. I think it would be great if we had something like `default_pic=true` for openssl 1.1. it would be a big ergonomics win.

winksaville commented:
I'll work on it this weekend.

winksaville commented:
I needed to suppress the unused function error when default_openssl wasn't being used.

dipinhora commented on Makefile:
```diff
@@ -859,10 +869,10 @@ benchmark: all
 
 stdlib-debug: all
-	$(PONY_BUILD_DIR)/ponyc -d --checktree --verify packages/stdlib
+	$(PONY_BUILD_DIR)/ponyc $(OPENSSL) -d --checktree --verify packages/stdlib
```
The `$(OPENSSL)` argument to `ponyc` should not longer be required due to the use of `default_openssl`.

winksaville replied:
The use of default_openssl is optional so i believe these are still needed.

dipinhora replied:
I lean towards recommending people use `default_openssl` similar to `default_pic` and removing these extra arguments but I don't feel very strongly about it.

I leave it to you and other other folks who decide to weigh in to make a decision about what to do.

winksaville replied:
I agree, its not needed if you assume you don't want change the value when using the Makefile. But one use I had was to use the Makefile to build the compiler and with a target of test-stdlib using default_openssl=openssl_1.1.0. This would compile the compiler and stdlib and then run it. I can then use the Makefile again to compile only stdlib using OPENSSL=-Dopenssl_0.9.0 to test that stdlib would not compile.

Probably not a normal case, but I have used it. I don't have a strong opinion either, but given that this is working and has some uses I'm inclined to leave.

dipinhora commented on src/libponyc/pkg/buildflagset.c:
```diff
@@ -534,6 +534,35 @@ bool define_build_flag(const char* name)
+  pony_assert(flags != NULL);
+
+  if(_user_flags == NULL)
+  {
```
It seems as though you should be able to `return false;` here if `_user_flags == NULL` since that means no user flags have been defined.

winksaville replied:
Yea, I actually coded it that way, but then thought its possible the initialization could potentially add flags and thought it best not to assume anything and just fall through.

dipinhora replied:
No, initialization doesn't add any flags. All it does is initialize the hashmap. See: https://github.com/ponylang/ponyc/blob/master/src/libponyrt/ds/hash.c#L180-L202 for what `flagtab_init` does.

Note: I should clarify that the way the hashmap implementation works there's a bunch of template macros which effectively alias the hashmap functions for a given type/name. If you find a `DECLARE_HASHMAP(flagtab, flagtab_t, flag_t);` with the same prefix then that is what is happening and you can look up the relevant hashmap functions to see what is happening if you need to.

winksaville replied:
I absolutely agree that now we could return early, but I see no downside to letting the normal processing proceed. I also tend to like code that has only and one exit point.

dipinhora replied:
Okey, in that case you can change this logic to be `if(_user_flags != NULL) { for... }` and keep a single exit point. Or you can keep things as they are. Like the other item, I don't feel strongly about it........

winksaville replied:
Then unless overriden by higher powers, I'll leave it as is :)

dipinhora commented on src/libponyc/pkg/buildflagset.c:
```diff
@@ -534,6 +534,35 @@ bool define_build_flag(const char* name)
+  {
+    flag_t f1 = {stringtab(*next), false};
+    size_t index = HASHMAP_UNKNOWN;
+    flag_t* f2 = flagtab_get(_user_flags, &f1, &index);
```
You can replace the `flagtab_get` and `flagtab_removeindex` combination with `flagtab_remove` instead (see here for header definition of `ponyint_hashmap_remove` function: https://github.com/ponylang/ponyc/blob/master/src/libponyrt/ds/hash.h#L83-L88). `flagtab_remove` will return `NULL` if the element wasn't found or else will return the element.  Please also remember to free the element after removing it to avoid a memory leak.

winksaville replied:
ok

dipinhora commented on src/libponyc/pkg/buildflagset.c:
```diff
@@ -534,6 +534,29 @@ bool define_build_flag(const char* name)
+  for(const char** next = flags; *next != NULL; next += 1)
+  {
+    flag_t f1 = {stringtab(*next), false};
+    if(flagtab_remove(_user_flags, &f1) != NULL)
```
You need to save the return `flag_t*` from `flagtab_remove` and then free it via `flag_free` if it is not `NULL` to avoid a memory leak.

winksaville replied:
Good catch, fixing.

dipinhora commented on src/libponyc/options/options.c:
```diff
@@ -242,7 +300,23 @@ ponyc_opt_process_t ponyc_opt_process(opt_state_t* s, pass_opt_t* opt,
+          if(validate_openssl_flag(s->arg_val))
+          { // User wants to add an openssl_flag,
+            // remove any existing openssl_flags.
+            remove_build_flags(valid_openssl_flags);
```
This line can be wrapped in an `#if defined(PONY_DEFAULT_OPENSSL)` since there wouldn't be a default openssl build flag to remove otherwise.

winksaville replied:
I'm in of favor minimizing assumptions and not adding conditionals if they aren't needed. So I'd rather leave it as is.

dipinhora replied:
Okey dokey. Works for me.

dipinhora commented on Makefile:
```diff
@@ -292,6 +292,26 @@ ifeq ($(runtime-bitcode),yes)
+  ifeq (ok,$(OPENSSL_VALID))
+    $(warning targeting openssl $(OPENSSL))
+  else
+    $(error OPENSSL=$(OPENSSL) is invalid, expecting one of -Dopenssl_0.9.0 or -Dopenssl_1.1.0)
```
This is likely to be confusing if someone specifies an invalid `default_openssl` value but the error message says that `OPENSSL` is invalid.

winksaville replied:
Good point, fixing.

winksaville commented:
@dipinhora et al. I've made some additional changes. First, I slept on the discussion with Dipin about OPENSSL and decided it should be removed. And second, did some cleanup of default pic and openssl handling, I think the changes improve the PR but your opinions may differ :)

Please review again.

winksaville commented:
I created an [issue for updating the FAQ](https://github.com/ponylang/ponylang.github.io/issues/234) when this is merged or another solution is found.

SeanTAllen commented:
This looks good. @winksaville is going to let me know when he opens a PR to update the FAQ on the website and i'll merge both at the same time.

winksaville commented:
I've got [PR 235](https://github.com/ponylang/ponylang.github.io/pull/235) for the FAQ.
